$ The canonical format for Linux kernel patches

This format is defined in `→ Linux/kernel/source/Documentation/process/submitting-patches_rst`.

A patch consists of
  • its *changelog* and
  • a marker line consisting of three hyphens (`---`)
  • optional comments (if used followed by an empty line)
  • the *patch* itself (in unified diff format)

These two parts are separated by `---` line.


{ Change log

  The format of the change log is
    • `from` line (optional, names the author of the patch if different from the person sending the patch)
    • One-line summary
    • A blank line
    • Detailed description
    • One or more tag lines

  Possible tag lines include:
    • `Signed-off-by:` (required)
    • `Fixes:` (First 12 characters of the commit id where problem was introduced plus its one-line summary in `("…")` format)
    • `Link:` (URL plus description)
    • `Closes:` (URL of bug report plus description)
    • `tag:`
    • `Co-developed-by:`
    • `Acked-by:`
    • `Tested-by:`
    • `Reviewed-by:`
    • `Reported-by:`
    • `Cc:`


}
{ Optional comments

  The optional comments don't go into the changelog.
  -
  Such comments might be used to describe the changes from the previous patch if an unaccepted
  patch is amended and resubmitted with a new version number (`[PATCH v2]…`).

  The optional comments might also include the output as produced by `diffstat -p 1 -w 70`.

}
{ Patch

  The patch itself comes in unified (= `→ Linux/shell/commands/diff -u`) format.
  -
  Adding `-p` (associated function names with changes) makes it easier for reviewers to read.

}


{ Message body

  The message body looks like
code {
from: John Doe <jd@somewhere.xyz>

A commit message. This message will be added to the permanent changelog
and describes the patch. It is separated by an empty line from the
preceding from: line.

The from line is only needed if the person sending the patch is different
from the sender's email address. This person will be credited as the
author of the patch.

The message is followed by one or more "signed off" lines and possibly other
tag lines which then is followed by a line containing three hyphens.

Signed-off-by: Donald Duck <dd@disneyland.com>
---
The optional text after the three hyphens won't go into the changelog.
It can be used for additional comments, diffstat etc.
It is followed by the actual diff output:

  arch/arm/boot/dts/keystone.dtsi | 8 ++++++++
  1 file changed, 8 insertions(+)

diff --git a/arch/arm/boot/dts/keystone.dtsi b/arch/arm/boot/dts/keystone.dtsi
index 3f272826f537..007648971744 100644
--- a/arch/arm/boot/dts/keystone.dtsi
--- b/arch/arm/boot/dts/keystone.dtsi

   etc
   etc
--
2.1.0
code }

  The *summarizing phrase* from the *Subject* becomes a globally-unique
  identitifier for that patch. It should not be longer than 75 characters.

}
{ Sending the patch

  The patch is sent to the corresponding → Linux/kernel/source/MAINTAINERS[maintainers] with such a *Subject* format: `[PATCH] subsystem: summarizing phrase`.

}
sa:

 `→ Linux/kernel/source/Documentation/email-clients_txt`

  The → Linux/kernel/source/scripts[scripts]:
   • `→ Linux/kernel/source/scripts/get_maintainer_pl`
   • `→ Linux/kernel/source/scripts/checkpatch_pl`

links:

  → http://linux.yyz.us/patch-format.html

  → https://www.ozlabs.org/~akpm/stuff/tpp.txt

  → https://lkml.org/lkml/2005/4/7/183
